### ISSUE TITLE
Masking with BiDirectional LSTMs not working?

### ISSUE THREAD

**Issue by hfjn** [keras-team/keras#7244]
Hi,

I'm trying to implement a nonlinear TimeSeries Prediction similar to this publication: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5336098/
Though I'm not interested in using a CNN as the first layer. I have a dataset of this shape:
`(249, None, 55)`, which I padded with 0s to have this shape: `(249, 543, 55)`. This works just fine with normal LSTMs.

For Bidirectional LSTM I implemented my model like this:
```
model = Sequential()
model.add(Masking(mask_value=0., input_shape=(None, X.shape[2])))
model.add(Bidirectional(LSTM(1000, return_sequences=True)))
model.add(Dropout(0.5))
model.add(Bidirectional(LSTM(1000, return_sequences=True)))
model.add(TimeDistributed(Dense(1000, activation="linear")))
model.add(Dropout(0.5))
model.add(TimeDistributed(Dense(1, activation="linear")))

model.compile(loss="mse", optimizer="adam", metrics=["mae"])
```
Unfortunately I'm getting weird results with this model which look like the model is partly ignoring the masking.

The problem with BiDirectional LSTMs and masking seems to be that by reading that data backwards, masking is once in the front (where it should be) and once in the back (where it shouldn't). Therefore I'm getting weird results. I found this old [blog post](http://dirko.github.io/Bidirectional-LSTMs-with-Keras/) which handles a similar problem though it using the old API. 

Do I need to provide distinct datasets with the padding in the correct place for each layer of the Bi-LSTM? And how would I feed that data into the model? Or should I leave the padding out completely and create batches of same size? Or even use `batch_size=1`? 

- [x] Check that you are up-to-date with the master branch of Keras. You can update with:
pip install git+git://github.com/fchollet/keras.git --upgrade --no-deps

- [x] If running on TensorFlow, check that you are up-to-date with the latest version. The installation instructions can be found [here](https://www.tensorflow.org/get_started/os_setup).

Cheers,

hfjn

**Comment by Cerno-b** [comment]
I have the same problem. Did you find out what it was? I tried setting the batch size to 1 but the test set accuracy was pretty bad (didn't try hyperparameter tuning, e.g. learning rate, so it may be my mistake).

One aspect I was wondering about: You said, masking would expect the masking to be in the front, not in the back. Since I don't know the internals about what Keras/TF does with masking exactly, it seemed a bit counterintuitive to me. Since I would expect a lot of sequence data has its padding at the end, I would expect the masking to be at the end, not in front.

Another issue I am unsure about with Bidirectional LSTMs with masking is whether the reverse function is working correctly. I looked into the code and it seems like the reverse function simply mirrors the data, which moves the padding from the back to the front as you wrote. However, I know that there is a TF function that does intelligent reversal which mirrors the data itself but keeps the padding at the end. However this function is not used by Keras, so I am wondering whether naive reverse with masking does the right thing.

**Comment by hfjn** [comment]
@Cerno-b Unfortunately, I never solved this problem. As I was working under strict time constraints (Did this for master thesis) I instead implemented a Deep LSTM for my problem. However, I agree that the reversal of padded strings in Keras is likely the origin of this problem and a way of "intelligently" reversing them would likely solve it.

**Comment by marchss** [comment]
is this issue solved? it sounds to be very serious as Bidirectional layer is commonly used.
